Feature/prplwrt build with a script

rmelotte opened the pull request:
Openwrt: move the build to a separate script, and add an option to build an "openwrt-builder" image only.

Moving the build part to a script makes it more readable, and easier
to maintain.

It also makes it easier to repeat the same steps in a container
interactively, to debug a failing build for example.

As a bonus, it makes it possible to run shellcheck and formatting
tools on it.

Some developers like to build OpenWrt interactively, to configure
different build options for example, or to be able to debug failing
builds more easily than what can be done when building a docker image.

With these changes, it's now possible to build the "openwrt-builder"
image:

tools/docker/builder/openwrt/build.sh --docker-target-stage
openwrt-builder --target-device netgear-rax40

The image can then be used to compile OpenWrt interactively.

rmelotte committed:
tools: openwrt: move the actual build to a separate script

Moving the build part to a script makes it more readable, and easier
to maintain.

It also makes it easier to repeat the same steps in a container
interactively, to debug a failing build for example.

As a bonus, it makes it possible to run shellcheck and formatting
tools on it.

- Move the last "RUN" step to build-openwrt.sh.

- Export the build arguments as environment variables to make them
  available to the script at build time, and to also keep them in the
  environment of the containers.

- Copy build-openwrt.sh into the container.

- Run build-openwrt.sh as the last RUN step.

- While we're at it, add license headers to the files we modify.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
tools: openwrt: add an option to build an openwrt-builder image

Some developers like to build OpenWrt interactively, to configure
different build options for example, or to be able to debug failing
builds more easily than what can be done when building a docker image.

tools/docker/builder/openwrt/Dockerfile:

- Add a new target, "openwrt-builder", that stops right before openwrt
  is actually built.

tools/docker/builder/openwrt/build.sh:

- Add a new option: "--docker-target-stage". When the option is used,
  the script will build the docker image only, and will stop at the
  specified target stage.

- Use the content of DOCKER_TARGET_STAGE for the default tag.

- Add the missing license header.

With these changes, it's now possible to build the "openwrt-builder"
image:

tools/docker/builder/openwrt/build.sh --docker-target-stage
openwrt-builder --target-device netgear-rax40

The image can then be used to compile OpenWrt interactively.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

tomereli commented on tools/docker/builder/openwrt/Dockerfile:
```diff
@@ -33,29 +40,38 @@ USER openwrt
 #
 # OpenWrt repository to use. Can also be a prplWrt repository:
 ARG OPENWRT_REPOSITORY
+ENV OPENWRT_REPOSITORY=$OPENWRT_REPOSITORY
```
Will this set an environment variable available in the container?

rmelotte replied:
Yes exactly, it makes them also available at run time.

arnout reviewed (approved):
Instead of adding the --docker-target-stage option (which is a bit annoying as an open-ended option, because it really can only be openwrt-prerequisites, openwrt-builder or prplmesh-builder), why don't we just tag all three of the, and change the --tag option to set only the part behind the colon?

That is however something that can be done in a separate commit/PR.

rmelotte committed:
tools: openwrt: output what is being done

The grep filter in .gitlab-ci.yml is also not needed anymore, since we
don't run make with with V=sc anymore.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>
Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

rmelotte commented:
@arnout Good idea. Since merging this is not urgent, I'll leave it open for now and update it.

rmelotte commented:
I have some more changes to make that will need to be reviewed anyway, so in the end I will merge this one first and open a new PR.
